src/llvm_ir_generator.cpp

```diff
@@ -0,0 +1,807 @@
+  llvm_util_.CreateBrIfNoBrBefore(step_bb);
+
+  builder_->SetInsertPoint(step_bb);
+  for_stmt.step->Accept(*this);
+  builder_->CreateBr(pred_bb);
+  builder_->SetInsertPoint(end_bb);
+}
+
+void LLVMIRGenerator::Visit(const ReturnStmtNode& ret_stmt) {
+  ret_stmt.expr->Accept(*this);
+  auto expr = val_recorder.ValOfPrevExpr();
+  builder_->CreateRet(expr);
+}
+
+void LLVMIRGenerator::Visit(const GotoStmtNode& goto_stmt) {
+  llvm::BasicBlock* target_bb = llvm_util_.FindBBWithNameOf(goto_stmt.label);
+
+  if (target_bb) {
+    builder_->CreateBr(target_bb);
+  } else {
+    auto label_bb = llvm::BasicBlock::Create(*context_, goto_stmt.label,
+                                             llvm_util_.CurrFunc());
+    builder_->CreateBr(label_bb);
+  }
+}
+
+void LLVMIRGenerator::Visit(const BreakStmtNode& break_stmt) {
+  assert(!label_views_of_jumpable_blocks.empty());
+  llvm_util_.CreateBrIfNoBrBefore(label_views_of_jumpable_blocks.back().exit);
+}
+
+void LLVMIRGenerator::Visit(const ContinueStmtNode& continue_stmt) {
+  assert(!label_views_of_jumpable_blocks.empty());
+  llvm_util_.CreateBrIfNoBrBefore(label_views_of_jumpable_blocks.back().entry);
+}
+
+void LLVMIRGenerator::Visit(const SwitchStmtNode& switch_stmt) {
+  switch_stmt.ctrl->Accept(*this);
+  auto ctrl = val_recorder.ValOfPrevExpr();
+  auto end_bb =
+      llvm::BasicBlock::Create(*context_, "switch_end", llvm_util_.CurrFunc());
+  auto sw = builder_->CreateSwitch(ctrl, nullptr);
+  label_views_of_jumpable_blocks.push_back({.entry = end_bb, .exit = end_bb});
+  switch_stmt.stmt->Accept(*this);
+  // Update cases and default label.
+  auto switch_infos = label_views_of_jumpable_blocks.back();
+  for (auto i = std::size_t{0}, e = switch_infos.cases.size(); i < e; ++i) {
+    auto case_info = switch_infos.cases.at(i);
+    auto curr_bb = case_info.second;
+    if (!case_info.first) {  // default case
+      sw->setDefaultDest(curr_bb);
+    } else {
+      auto const_val = llvm::dyn_cast<llvm::ConstantInt>(case_info.first);
+      assert(const_val);
+      sw->addCase(const_val, curr_bb);
+    }
+
+    // NOTE: If BB has no terminator and has a next BB, add one branch to next
+    // BB. If BB is the last BB, then branch to switch exit.
+    if (i + 1 != e) {
+      auto next_bb = switch_infos.cases.at(i + 1).second;
+      llvm_util_.CurrBBFallThroughNextBB(curr_bb, next_bb);
+    } else {
+      llvm_util_.CurrBBFallThroughNextBB(curr_bb, switch_infos.exit);
+    }
+  }
+  label_views_of_jumpable_blocks.pop_back();
+  llvm_util_.CreateBrIfNoBrBefore(end_bb);
+  builder_->SetInsertPoint(end_bb);
+}
+
+void LLVMIRGenerator::Visit(const IdLabeledStmtNode& id_labeled_stmt) {
+  llvm::BasicBlock* target_bb =
+      llvm_util_.FindBBWithNameOf(id_labeled_stmt.label);
+
+  if (!target_bb) {
+    auto label_bb = llvm::BasicBlock::Create(*context_, id_labeled_stmt.label,
+                                             llvm_util_.CurrFunc());
+    builder_->SetInsertPoint(label_bb);
```

Comment:
Similarly, if we are creating the basic block here when the target doesn't exist yet, we can reassign the value to `target_bb` instead of declaring a new one. This better conveys the fact that we are creating the block.